mpremote: Implement mip install from GitLab.

OlivierLenoir opened the pull request:
mpremote: Implement mip install from GitLab.

tools/mpremote/README.md: Add mip install gitlab:org/repo@branch.
Add example `mpremote mip install gitlab:org/repo@branch`.

tools/mpremote/mpremote/mip.py: Implement install from GitLab.
Modify _rewrite_url() to allow mip install from `gitlab:` repository.

mpremote/mpremote/main.py: Add gitlab:org/repo, gitlab:org/repo@branch.
Modify argparse_mip() for gitlab:org/repo, gitlab:org/repo@branch.

docs/reference/packages.rst: Mip install from GitLab.
Describe how to mip install from GitLab.

docs/reference/mpremote.rst: Install package from GitLab.
Install the package from the specified branch at org/repo on GitLab.

Signed-off-by: Olivier Lenoir <olivier.len02@gmail.com>

OlivierLenoir commented:
Both pull requests are link to the same implementation of mip intall from GitLab.

- https://github.com/micropython/micropython-lib/pull/816
- https://github.com/micropython/micropython/pull/13791

Signed-off-by: Olivier Lenoir olivier.len02@gmail.com

-------------------------------------------------------
https://github.com/micropython/micropython-lib/pull/816

Modify _rewrite_url() to allow mip install from gitlab: repository.

Signed-off-by: Olivier Lenoir olivier.len02@gmail.com

-----------------------------------------------------
https://github.com/micropython/micropython/pull/13791

mpremote: Implement mip install from GitLab.

tools/mpremote/README.md: Add mip install gitlab:org/repo@branch.
Add example mpremote mip install gitlab:org/repo@branch.

tools/mpremote/mpremote/mip.py: Implement install from GitLab.
Modify _rewrite_url() to allow mip install from gitlab: repository.

mpremote/mpremote/main.py: Add gitlab:org/repo, gitlab:org/repo@branch.
Modify argparse_mip() for gitlab:org/repo, gitlab:org/repo@branch.

docs/reference/packages.rst: Mip install from GitLab.
Describe how to mip install from GitLab.

docs/reference/mpremote.rst: Install package from GitLab.
Install the package from the specified branch at org/repo on GitLab.

Signed-off-by: Olivier Lenoir olivier.len02@gmail.com

jimmo commented on docs/reference/packages.rst:
```diff
@@ -56,6 +56,24 @@ If no json file is specified, then "package.json" is implicitly added::
     >>> mip.install("github:org/repo")  # Uses default branch of that repo
     >>> mip.install("github:org/repo", version="branch-or-tag")
 
+The URL can also start with ``gitlab:`` as a simple way of pointing to content
```
Thanks for adding docs.

This section mostly just duplicates the above though, and because the gitlab support works in exactly the same way at github, maybe just say

```
The URL can also start with ``gitlab:`` and works in exactly the same way as GitHub URLs described above ``gitlab:org/repo[/path]``, including support for branches.
```

jimmo reviewed (commented):
Looks good, thanks @OlivierLenoir !

OlivierLenoir commented:
@jimmo , I've aggregated both topics into one. I've also added a _deps_ example with _gitlab:_.
Let me know if you valid the documentations modification.

jimmo commented on docs/reference/packages.rst:
```diff
@@ -38,24 +38,28 @@ install third-party libraries. The simplest way is to download a file directly::
-The URL can also start with ``github:`` as a simple way of pointing to content
-hosted on GitHub::
+The URL can also start with ``github:`` or ``gitlab:`` as a simple way of pointing to content
+hosted on GitHub:: or GitLab::
```
Extra `::` after `GitHub`

jimmo reviewed (commented):
Thanks for updating! Please squash the commits down to a single commit (or if you want, one to the code and one to the docs).

OlivierLenoir commented:
@jimmo, I've fix the extra :: after github.
Regarding your demand to squash down to a single commit, I'm not certain how to do, it's my first real contribution to MicroPython project.
I plan to rebase my local git repository, but is it the right method?
```shell
git rebase -i HEAD~<number_of_commit>
git push --force-with-lease
``` 
As 5 files have been updated, what should be the commit first line and comment?
Can you help?

OlivierLenoir committed:
tools/mpremote: Add support to mip install from GitLab.

Changes are:
- main.py: Add gitlab:org/repo, gitlab:org/repo@branch.
- mip.py: Implement install from GitLab.
- README.md: Add mip install gitlab:org/repo@branch example.

Signed-off-by: Olivier Lenoir <olivier.len02@gmail.com>

OlivierLenoir committed:
docs/reference: Document how to mip install packages from GitLab.

Signed-off-by: Olivier Lenoir <olivier.len02@gmail.com>

robert-hh commented:
You PR seems completely messed up, since it contains all the STATIC->static changes.
Did you rebase your branch before changing the STATIC case locally? Then

OlivierLenoir commented:
> You PR seems completely messed up, since it contains all the STATIC->static changes. Did you rebase your branch before changing the STATIC case locally? Then

I'm trying to rebase my commit into one, but it fail because I did not format the commit message as it should.
It seem that I'm not doing it the right way.
@robert-hh, what should I do? Drop my fork and start again? Thanks for your help.

jimmo commented:
@OlivierLenoir This might help: https://github.com/jimmo/git-and-micropython

Basically if you do a `rebase -i` and remove all the lines for commits that aren't relevant to your PR.

OlivierLenoir commented:
@jimmo I think it's now correct, thanks to your help: [https://github.com/jimmo/git-and-micropython](https://github.com/jimmo/git-and-micropython).
What is the next step, if any?

stinos commented on docs/reference/packages.rst:
```diff
@@ -38,24 +38,28 @@ install third-party libraries. The simplest way is to download a file directly::
 
     >>> mip.install("http://example.com/x/package.json")
     >>> mip.install("github:org/user/path/package.json")
+    >>> mip.install("gitlab:org/user/path/package.json")
```
Just wondering but do we really want to repeat this in every single example in all involved readme/.rst files? Because suppose we add another method then we'd also have to adjust all locations again?

OlivierLenoir replied:
I just consider that example should be explicit to ease search and understanding.
I've done it searching for _github:_, and now _gitlab:_.
```shell
$ grep -rl "github:\|gitlab:"
tools/mpremote/README.md
tools/mpremote/mpremote/mip.py
tools/mpremote/mpremote/main.py
docs/reference/packages.rst
docs/reference/mpremote.rst
```

dpgeorge commented:
Thanks for the contribution!  Now merged.
